Legacy LogConsumer provider sink for Microsoft.Extensions.Logging

xiazen opened the pull request:
Added `OrleansLogger `and `OlreansLoggerProvider` to preserve legacy Orleans logging features and configuration style, including

- Message Bulking, including its configuration
- `ILogConsumer`, `IClosableLogConsumer`, `IFlushableLogConsumer`, including its configuration
- per category Severity Overrides, whose configuration style preserves old severity config style

Added OrleansLoggerExtension methods to preserve old Orleans logging methods. 

Wired asp.net logging abstraction into silo's DI container, with OrleansLoggerProvider as default logger provider, LogFileConsumer as default ILogConsumer. One can now configure his LogFactory into DI, one can also create ILogger<T> from DI now. See tests as example.

`OrleansLoggerProvider` supports
- Create `new OrleansLoggingDecorator(bulkingConfig, OrleansLogger logger)`
- Configure LogConsumers
- Configure Severity overrides
- Configure message bulking

`OrleansLoggingDecorator` 
- a decorator class which user can plug in their ILogger and then gain message bulking feature on top of their ILogger

`OrleansLogger` supports
- log messages to configured LogConsumers
- respects Severity overrides in `IsEnable(LogLevel level)` method
- Map LogLevel to Severity

`FileLogConsumer` is the replacement for FileTelemetryConsumer, which also preserves Orleans logging message format. Since general consensus I get is TelemetryConsumer should be replaced by asp.net built in log providers, or at least separate out from OrleansLogging. 

`ILogger `and `OlreansLoggerProvider` is supposed to replace `Logger`, `LogManager` eventually. OrleasnLogger is supposed to replace LoggerImpl. I haven't changed them  in this PR. So they work as is for now. 

About how `Microsoft.Extensions.Logging.LoggingFilter` and `OrleansLoggerSeverityOverrides` play together:
One can use either `Microsoft.Extensions.Logging.LoggingFilter` or `OrleansLoggerSeverityOverrides` to configure per logger Severity/LogLevel. Both config method works with `OrleansLogger`. If used together, think it as two layer of filters, need both filter being true for the end result to be true. 
See tests to get a clearer idea

EDIT:
- TelemetryConsumers are not wired into OrleansLogging eco system. How it would play into the picture is unclear, since there's different opinion about it. Need to get back to that discussion.  **EDIT:** I have received general consensus that Telemetry will evolve to have its own system, which doesn't share infrastructure with logging. So remove this TODO from this PR. This will be resolved in its own path with its own series of PRs. 
- support logging scope. **EDIT: ** Can be done in a separate PR. It wasn't part of Orleans legacy logging feature. So merge this PR without this piece won't break anyone. But need to be added before 2.0
- Move logging configuration to SiloBuilder. **EDIT:**  logging can be configured with DI. So I don't think there need to be extra config method on SiloBuilder for configuring logging. Although there can be that kind of config method on SiloBuilder, but I think it is out of scope of this PR. So remove this TODO from this PR.

xiazen commented on src/Orleans/Logging/ILogConsumer.cs:
```diff
@@ -47,19 +47,15 @@ public interface ILogConsumer
         /// In general, all log entries at severity=Error or greater should specify an explicit error code value.</param>
         void Log(
             Severity severity,
-            LoggerType loggerType,
```
We agreed to remove LoggerType in that meeting. Is this correct? 

If LoggerType is removed from Orleans logging infra, its downstream ILogConsumer will not be able to know the LoggerType.

xiazen commented on src/Orleans/Telemetry/Severity.cs:
```diff
@@ -4,10 +4,11 @@
     public enum Severity
     {
         Off = 0,
-        Error = 1,
```
it changed to have a better mapping to LogLevel. But yeah I think I should change it back. it's not worthwhile

xiazen commented on src/Orleans/Logging/LoggerImpl.cs:
```diff
@@ -240,13 +240,12 @@ private void WriteLogMessage(int errorCode, Severity sev, string format, object[
                 try
                 {
-                    consumer.Log(sev, loggerType, Name, message, LogManager.MyIPEndPoint, exception, errorCode);
+                    consumer.Log(sev, Name, message, exception, errorCode);
```
Yes that's correct. I Haven't made that change in this PR. Orleans.Logger will become a wrapper on top of ASP.Net ILogger . Is this what you are asking?

xiazen replied:
The following PR will be replacing Orleans legacy infra with new ASP.net abstraction

jdom commented:
Haven't started reviewing, althogh from the description it looks like we are deprecating stuff and calling it `OrleansXXX` (ie OrleansLogger) and similar. I would call it `LegacyXXX` instead, otherwise it prompts users as the way to go for future code. Thoughts?

jdom commented on vNext/src/Orleans/Orleans.csproj:
```diff
@@ -33,6 +33,8 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="1.1.1" />
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="2.0.0-preview2-final" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.0.0-preview2-final" />
```
Logging 2.0.0 was released. Although it's a .NET Standard 2.0 library, so let's figure this out before we even try to merge

galvesribeiro replied:
It works. But be aware that only people with latest VS/Nuget will be able to add `netstandard2.0` libraries on .Net Full framework projects.

xiazen replied:
Yeah I need to update the PR with Logging 2.0.0. The PR is created before that was released.

xiazen replied:
Yes this PR is not mergeable yet. There's some TODOs(see description) which need to be addressed before merging. I'd like to do some of there todos after the two sln merged into one sln.

xiazen commented:
I was thinking adding `[System.Obsolete] `attribute on those classes.  Do you think it is not enough?

xiazen commented on vNext/src/Orleans/OrleansLogging/FileLogConsumer.cs:
```diff
@@ -0,0 +1,85 @@
+    /// <summary>
+    /// FileLogConsumer, which logs message into a file in orleans logging message style
+    /// </summary>
+    public class FileLogConsumer : ICloseableLogConsumer, IFlushableLogConsumer
```
`FileLogConsumer` is the replacement for `FileTelemetryConsumer`. FileTelemetryConsumer is the default component in Orleans to record tracing messages. But since we are separating logging and telemetry, so FileTelemetryConsumer won't be compatible with asp.net logging abstraction anymore. I created FileLogConsumer to replace its role in logging, just in case that people still want to log messages in Orleans logging message style.

xiazen commented:
ready for review

xiazen commented:
@dotnet-bot test netfx-functional please

xiazen commented on src/OrleansLoggerProvider/OrleansLoggerProvider.csproj:
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
```
ILogConsumer and all the deprecated logging features will be moved here as much as possible when internal runtime finished migrating off the legacy logging infra

xiazen commented on src/OrleansLoggerProvider/OrleansLoggingDecorator.cs:
```diff
@@ -74,9 +74,9 @@ public OrleansLoggingDecorator(MessageBulkingConfig config, ILogger thirdPartyLo
         /// <param name="eventId"></param>
         /// <param name="errorCode"></param>
         /// <returns></returns>
-        public static EventId CreateEventId(int eventId, int errorCode)
```
this will be replaced by LoggingUtils.CreateEventId n #3333 after it merged

ReubenBond commented on src/OrleansLoggerProvider/OrleansLogger.cs:
```diff
@@ -0,0 +1,119 @@
+            if ((DateTime.UtcNow - lastFlush) > flushInterval)
+            {
+                lastFlush = DateTime.UtcNow;
+                foreach (IFlushableLogConsumer consumer in this.logConsumers.OfType<IFlushableLogConsumer>())
```
The `this.logConsumers.OfType<IFlushableLogConsumer>()` could be computed once and enumerated to a list in the ctor.

ReubenBond commented on src/OrleansRuntime/Silo/Silo.cs:
```diff
@@ -886,6 +889,16 @@ private void HandleProcessExit(object sender, EventArgs e)
 
+        private void DisposeLogProviders(IServiceProvider svc)
+        {
+            var providers = svc.GetServices<ILoggerProvider>();
```
This shouldn't be required - the container is disposed immediately after this call and it is responsible for disposing these.

xiazen replied:
no it isn't disposing the providers. I tested it.

xiazen replied:
nvm. you are right. I was doing things wrong

ReubenBond commented on src/OrleansLoggerProvider/OrleansLoggerProvider.cs:
```diff
@@ -0,0 +1,155 @@
+        /// <returns></returns>
+        public Severity FindSeverityLevel(string categoryName)
+        {
+            if (this.loggerSeverityOverrides.LoggerSeverityOverrides.ContainsKey(categoryName))
```
better to use `TryGetValue` here

xiazen commented:
feedback I received online or offline require changes which is blocked by #3368 . Will go back after #3368 is merged

xiazen commented:
ready for review

xiazen commented:
please review this, and hopefully can merge this soon. Since this PR adds a new csproj, so the longer we delay its merging, the more I need to deal with changes from project restructuring #3353 .

jdom commented:
Anyone has thoughts on [this comment](https://github.com/dotnet/orleans/pull/3284#issuecomment-322617798)? It doesn't feel right that we are *adding* OrleansLogger and it's already obsolete. IMO it should be LegacyLogger or even LegacyOrleansLogger

jdom commented on test/Tester/OrleansLoggingTests.cs:
```diff
@@ -0,0 +1,102 @@
+    public class OrleansLoggingTests
+    {
+        [Fact]
+        [Obsolete]
```
Why are these tests marked obsolete?

xiazen replied:
to avoid compile warnings

attilah replied:
Can't we pragma disable the warnings for the logging tests instead of putting on ```Obsolete``` on the test methods?

jdom commented on test/Tester/OrleansLoggingTests.cs:
```diff
@@ -0,0 +1,102 @@
+            Assert.False(logger.IsEnabled(LogLevel.Information));
+
+            //dispose log providers
+            (serviceProvider as IDisposable)?.Dispose();
```
instead of this line, wrap the `serviceProvider` into a `using()` statement. Same with other facts.

jdom commented on src/OrleansLoggerProvider/OrleansLoggingUtils.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <param name="myIPEndPoint"></param>
+        /// <param name="exception"></param>
+        /// <param name="errorCode"></param>
+        /// <param name="includeStackTrace"></param>
```
you copied this from the original code, why not include the XML comments?
```
        /// <summary>The method to call during logging to format the log info into a string ready for output.</summary>
        /// <param name="timestamp">Timestamp of the log message.</param>
        /// <param name="severity">The severity of the message being traced.</param>
        /// <param name="loggerType">The type of logger the message is being traced through.</param>
        /// <param name="caller">The name of the logger tracing the message.</param>
        /// <param name="myIPEndPoint">The <see cref="IPEndPoint"/> of the Orleans client/server if known. May be null.</param>
        /// <param name="message">The message to log.</param>
        /// <param name="exception">The exception to log. May be null.</param>
        /// <param name="errorCode">Numeric event code for this log entry. May be zero, meaning 'Unspecified'.</param>
```

galvesribeiro commented:
I'm with you @jdom.

jdom commented on src/OrleansLoggerProvider/OrleansLogger.cs:
```diff
@@ -0,0 +1,149 @@
+    /// OreansLogger supports legacy orleans logging features, including <see cref="ILogConsumer"/>, <see cref="ICloseableLogConsumer">,
+    /// <see cref="IFlushableLogConsumer">, <see cref="Severity">. 
+    /// </summary>
+    [Obsolete]
```
All obsolete attributes should have a description of what should be the replacement. You should probably define a constant with this string, as you will use it in multiple places.

jdom commented on src/OrleansLoggerProvider/OrleansLoggerProvider.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <summary>
+        /// LoggerSeverityOverrides, which key being logger category name, value being its overrided severity
+        /// </summary>
+        public ConcurrentDictionary<string, Severity> LoggerSeverityOverrides { get; }
```
just do `public ConcurrentDictionary<string, Severity> LoggerSeverityOverrides { get; } = new ConcurrentDictionary<string, Severity>();` and get rid of the constructor.
Event better, don't use a concurrent dictionary, as it is only used for configuration, just do:

`public IDictionary<string, Severity> LoggerSeverityOverrides { get; set; } = new Dictionary<string, Severity>();` 

This way users can set to their own dictionary instead of adding to this one.

jdom commented on src/OrleansRuntime/Hosting/DefaultSiloServices.cs:
```diff
@@ -55,6 +55,7 @@ internal static void AddDefaultServices(IServiceCollection services)
             services.TryAddFromExisting<IMessagingConfiguration, GlobalConfiguration>();
             services.TryAddFromExisting<ITraceConfiguration, NodeConfiguration>();
 
+            services.AddLogging();
```
technically this isn't needed nor used in this PR, right?

xiazen replied:
that's right. I can remove it

jdom commented on src/OrleansLoggerProvider/OrleansLoggingDecorator.cs:
```diff
@@ -0,0 +1,134 @@
+    /// OrleansLoggingDecorator class. User can plug in their own ILogger implementation into this decorator class to add message bulking feature on top of their logger. 
+    /// Message bulking feature will just log eventId count if the same eventId has appear more than BulkMessageLimit in a certain BulkMessageInterval.
+    /// </summary>
+    public class OrleansLoggingDecorator : ILogger
```
I'd rename to `MessageBulkingDecoratorLogger` or even (`EventBulkingDecoratorLogger`), as there is nothing inherent to Orleans nor the legacy logger. In theory this could be externalized as a separate package that non-Orleans users could use (not saying we should do the latter now though, just mentioning how decoupled this is).

xiazen replied:
will do

jdom commented on src/OrleansLoggerProvider/OrleansLoggingDecorator.cs:
```diff
@@ -0,0 +1,134 @@
+        /// </summary>
+        /// <param name="config"></param>
+        /// <param name="thirdPartyLogger"></param>
+        public OrleansLoggingDecorator(MessageBulkingConfig config, ILogger thirdPartyLogger)
```
please rename thirdPartyLogger to `decoratedLogger` everywhere (or `wrappedLogger` if you prefer), as 3rd party isn't correct.

xiazen replied:
will do

jdom commented on src/OrleansLoggerProvider/OrleansLoggingDecorator.cs:
```diff
@@ -0,0 +1,134 @@
+    /// <summary>
+    /// Config for message bulking feature
+    /// </summary>
+    public class MessageBulkingConfig
```
how about `MessageBulkingOptions` (or `EventBulkingOptions` if you also rename the decorator that way)?

xiazen replied:
I think `EventBulkingConfig` is good. Why do you prefer **options** over **config**, as in I think this is a config class?

jdom replied:
with the new builders, we are moving to using strongly typed Options (that could be bound using Microsoft.Extensions.Options). We'll get rid of our legacy Config and Configuration classes, and morph them to Options (meaning we'll remove any parsing logic from them too, something similar to this message bulking options)

jdom commented on src/OrleansLoggerProvider/OrleansLoggingFactoryExtension.cs:
```diff
@@ -0,0 +1,73 @@
+        public static ILoggingBuilder AddMesageBulkingLoggerProvider<TDecoratedLoggerProvider>(this ILoggingBuilder builder, TDecoratedLoggerProvider provider, MessageBulkingConfig messageBulkingConfig)
+            where TDecoratedLoggerProvider : ILoggerProvider
+        {
+            builder.AddProvider(new MessageBulkingLoggerProvider<TDecoratedLoggerProvider>(provider, messageBulkingConfig));
```
this should probably be conditional on whether the bulking options are not null. Otherwise just add the other provider directly.

xiazen replied:
I think the correct fix here is making bulkConfig optional, since it do have default config values. 
This extension method is designed to add event bulking feature onto the user's logger provider, so its behavior should stick with that IMO. 
If the user doesn't want event bulking feature, he/she can just use native builder.AddProvider(their logger provider).

jdom replied:
well, the point is that this is also being used by AddLegacyOrleansLogger... Then if not here, do the conditional there first and decide whether to call AddEventBulkingLoggerProvider whether that one is non-null

jdom commented on src/OrleansLoggerProvider/OrleansLoggingFactoryExtension.cs:
```diff
@@ -0,0 +1,73 @@
+        public static ILoggingBuilder AddOrleansLogging(
+            this ILoggingBuilder builder,
+            List<ILogConsumer> logConsumers,
+            IPEndPoint ipEndPoint = null,
```
hmmm, this ipEndPoint here might be tricky. We can probably address in a different PR, but maybe a solution is to add extension methods for ISiloBuilder and IClientBuilder also, so that we can resolve the IP endpoint first and then add the provider, but not sure about order of initialization.

xiazen replied:
we can discuss this in another PR. But IMO, this extensions should be decoupled with core interface as much as possible, since it is being deprecated. I see we provide convenience for users on how to resolve IpEndpoint using extenion method on ISiloBuilder, but at the mean time, we also couple those two closer than before, event it is just an extension method. I'm not concerned about resolving ipEndpoint, since it is pretty easy for user to resolve it as in they are the ones who configure that,.

jdom replied:
yes, let's discuss afterwards. Although it might not be as easy to know the ipendpoint though, as it might be resolved and bound by the silo itself, not the user

xiazen commented:
addressed feedbacks. added legacy keyword to all obsolete classes. Also added Orleans.Extensions.Logging.Legacy namespace, which included all obsolete classes.

jdom commented on test/Tester/OrleansLoggingTests.cs:
```diff
@@ -23,10 +24,10 @@ public void OrleansLoggingCanConfigurePerCategoryServeriyOverrides()
-                new FileLogConsumer($"{this.GetType().Name}.log")
+                new LegacyFileLogConsumer($"{this.GetType().Name}.log")
             }, severityOverrides));
             var serviceProvider = serviceCollection.BuildServiceProvider();
```
I meant that you do:
```
using (var serviceProvider = serviceCollection.BuildServiceProvider())
{
...
}
```
and then do not do not use Dispose explicitly. But on hindsight, BuildServiceProvider returns IServiceProvider which is not itself disposable, the concrete implementation is, so you might not be able to do it

xiazen replied:
no I cannot do that since serviceProvider is not disposable. you want me to revert to original line?

jdom replied:
Yes

jdom commented on src/Orleans/Orleans.csproj:
```diff
@@ -37,6 +37,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="$(MicrosoftExtensionsDependencyInjection)" />
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="$(MicrosoftExtensionsLoggingVersion)" />
```
please remove this one for now if it's not needed

xiazen replied:
it is needed for LegacyLogger project, since it  have a dependency for Microsoft.Extensions.Logging and also Orleans. And Orleans need to have dependency on Microsoft.Extensions.Logging anyways, I'd like to make this change as soon as possible.

jdom replied:
As long as we don't have the, let's not pollute things here, as this is a high impact change

jdom commented on src/OrleansLoggerProvider/LegacyOrleansLoggerProvider.csproj:
```diff
@@ -0,0 +1,18 @@
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <AssemblyName>LegacyOrleansLoggerProvider</AssemblyName>
```
Please exclude the redundant Orleans from the assembly and package name. And I'd remove Provider from there too, since we are probably going to add the Logger stuff too in the very near future.
Just call it LegacyLogging (and package name Microsoft.Orleans.LegacyLogging)

xiazen replied:
at a second thought, I think we should just name it as Logging and Microsoft.Orleans.Logging. Since this assembly includes both legacy feature and non-legacy feature, such as event bulking. 

I added Orleans.Extensions.Logging.Legacy namespace to differentiate legacy feature and non-legacy ones

jdom replied:
But bulking itself we are only including for legacy reasons. This should not be an Orleans concern, but an application concern. If we hadn't had that feature before, we should not be including this here.

attilah replied:
Agree, see my first comment in the review.

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLogger.cs:
```diff
@@ -0,0 +1,149 @@
+    /// LegacyOrleansLogger supports legacy orleans logging features, including <see cref="ILogConsumer"/>, <see cref="ICloseableLogConsumer">,
+    /// <see cref="IFlushableLogConsumer">, <see cref="Severity">. 
+    /// </summary>
+    [Obsolete("Use Microsoft.Extensions.Logging built-in loggers")]
```
I think we should rephrase it, because we encourage people to use ```ILogger``` calls to log everything but not to use builtin loggers, since they can use ANY conforming logger.

xiazen replied:
What do you think the words need to be? please use ILogger? it is kind of confusing, since LegacyOrleansLogger is an Ilogger. 

please tell me the words you think fit here

jdom replied:
How about "The Microsoft.Orleans.Logging.Legacy namespace was kept to facilitate migration from Orleans 1.x but will be removed in the near future. It is recommended that you use the Microsoft.Extensions.Logging infrastructure and providers directly instead of Microsoft.Orleans.Logging.Legacy.Logger and Microsoft.Orleans.Logging.Legacy.ILogConsumer."

(I understand you haven't migrated Logger yet, but you can skip that part for now).
Again, I'd put this in a constant, and refer to it from every Obsolete attribute in this namespace.

xiazen replied:
sounds good. except for it is kind of long lol

jdom replied:
Heh, but it's very explicit. BTW, you keep duplicating the string everywhere. Why are you not creating the constant and have the string defined just once?

xiazen replied:
hmmm..... insert texts explaining I know what i'm doing but in fact not.

xiazen replied:
fixed

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLogger.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <param name="maxSeverityLevel"></param>
+        public LegacyOrleansLogger(string categoryName, IList<ILogConsumer> logConsumers, IPEndPoint ipEndPoint)
+        {
+            this.logConsumers = logConsumers;
```
We only need an enumerable, but not modifiable version, I'd choose another type instead of ```IList```, same for ```flushableConsumers```.

Specially as we are calling it like this: ```this.LogConsumers.ToArray()```

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLogger.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="categoryName"></param>
```
No content for parameters.

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLogger.cs:
```diff
@@ -0,0 +1,149 @@
+        /// <inheritdoc/>
+        public bool IsEnabled(LogLevel logLevel)
+        {
+            return true;
```
Why no validating of log levels here?

xiazen replied:
no need, it is taken care by log filters

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLogger.cs:
```diff
@@ -0,0 +1,149 @@
+                case LogLevel.Error: return Severity.Error;
+                case LogLevel.Warning: return Severity.Warning;
+                case LogLevel.Information: return Severity.Info;
+                case LogLevel.Debug: return Severity.Verbose;
```
Since we cannot map ```Verbose2``` due to the lack of "enough" levels in the logging abstractions, IMHO we should replace all ```Verbose2``` calls with ```Verbose``` or ```Verbose3```.

For default I'd add at least a comment that it for ```Trace```, or explicitly make that a case (I prefer), and for default throw an exception (since that cannot happen, because we're mapping all values).

xiazen replied:
I believe this is not part of this PR's scope? and we are not mapping Verbose2 and Verbose3 to Trace, instead of "cannot map Verbose2"

xiazen replied:
it map LogLevel.Trace to Verbose3 here so that both Verbose2 and Verbose3 will be logged if user set default llogLevel to Trace.

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLogger.cs:
```diff
@@ -0,0 +1,149 @@
+                type = LoggerType.Runtime;
+                if (category.Contains("Provider"))
+                    type = LoggerType.Provider;
+                if (category.Contains("Grain"))
```
These can overwrite each other, no need for ```else``` ?

xiazen replied:
it is intended this way. it basically means, if category contains Orleans and Provider than it is LoggerType.provider

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLoggerProvider.cs:
```diff
@@ -0,0 +1,76 @@
+        /// </summary>
+        public static Severity DefaultSeverity = Severity.Info;
+
+        public ConcurrentBag<ILogConsumer> LogConsumers { get; private set; }
```
Why is it public? I see no external references to it.

xiazen replied:
it is one way user can configure log consumers... there's no external reference because we don't have user code here..

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLoggerProvider.cs:
```diff
@@ -0,0 +1,76 @@
+        /// <summary>
+        /// Default Severity for all loggers
+        /// </summary>
+        public static Severity DefaultSeverity = Severity.Info;
```
Why should we make this public?

xiazen replied:
so that user know what is the default Severity?

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLoggerProvider.csproj:
```diff
@@ -0,0 +1,18 @@
+  <PropertyGroup>
+    <AssemblyName>LegacyOrleansLoggerProvider</AssemblyName>
+    <RootNamespace>Orleans.Extensions.Logging</RootNamespace>
+    <PackageTargetFallback Condition=" '$(BuildFlavor)' == 'Legacy' ">$(PackageTargetFallback);portable-net45+win8+wpa81+wp8</PackageTargetFallback>
```
I don't think we need ```PackageTargetFallback``` here, since we're not referencing any NuGets that would require it.

xiazen replied:
cannot build without this line. may need more work here. can we address this in another PR. and get the bigger part out?

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLoggingFactoryExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+        [Obsolete("Use Microsoft.Extensions.Logging built-in logger providers")]
+        public static ILoggingBuilder AddLegacyOrleansLogging(
+            this ILoggingBuilder builder,
+            List<ILogConsumer> logConsumers,
```
```IEnumerable``` would be sufficent here.

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLoggingFactoryExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <param name="eventBulkingConfig">config for event bulking feature</param>
+        /// <returns></returns>
+        [Obsolete("Use Microsoft.Extensions.Logging built-in logger providers")]
+        public static ILoggingBuilder AddLegacyOrleansLogging(
```
This is an ```ILoggingBuilder``` extension class and not ```ILoggingFactory```, please add ```ILoggingFactory``` implementation. SeriLog can be a good sample here perhaps: https://github.com/serilog/serilog-extensions-logging/tree/dev/src/Serilog.Extensions.Logging

xiazen replied:
ILoggingBuilder is the 2.0 way of configuring logging. Serilog haven't migrated to it IMO because they haven't got to it. 

But take a look at Microsoft.Extensions.Logging.TraceSOurce code, they migrated to ILoggerBuilder. Let me @jdom here since he promoted that we migrated to ILoggerBuilder

attilah commented on src/OrleansLoggerProvider/LegacyOrleansLoggingFactoryExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+        {
+            if (builder == null)
+                throw new ArgumentNullException(nameof(builder));
+            builder.AddMesageBulkingLoggerProvider<LegacyOrleansLoggerProvider>(new LegacyOrleansLoggerProvider(logConsumers, ipEndPoint), eventBulkingConfig);
```
As a general rule of thumb, if extension method arguments differ, we should provide default values and forward the calls to the one method which does the work. From here it would be easy to not pass in ```severityOverrides``` and call the method below.

xiazen replied:
which general rule of thumb, maybe refer me a link? And I think as a team we should have an agreement on which coding guideline we are following, it provider convenience for the coder to know what style of code he/she should write, instead of discussing it in PRs

attilah commented on src/OrleansLoggerProvider/OrleansLoggingFactoryExtension.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Orleans.Extensions.Logging
+{
+    public static class OrleansLoggingFactoryExtension
```
This is an ```ILoggingBuilder``` extension class and not ```ILoggingFactory```, please add ```ILoggingFactory``` implementation. SeriLog can be a good sample here perhaps: https://github.com/serilog/serilog-extensions-logging/tree/dev/src/Serilog.Extensions.Logging

xiazen replied:
see my comment above

attilah commented on test/Tester/OrleansLoggingTests.cs:
```diff
@@ -0,0 +1,104 @@
+            {
+                new LegacyFileLogConsumer($"{this.GetType().Name}.log")
+            }, severityOverrides));
+            var serviceProvider = serviceCollection.BuildServiceProvider();
```
Why is DI involved in these tests? I see no need or value to it.

xiazen replied:
the runtime gets logger from DI. I want to emulate that

xiazen commented:
@dotnet-bot test netfx-bvt please

jdom commented on Orleans.sln:
```diff
@@ -118,7 +118,9 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OrleansGCPUtils", "src\Orle
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Orleans.Core.Abstractions", "src\Orleans.Core.Abstractions\Orleans.Core.Abstractions.csproj", "{73514686-D25D-478B-9943-A86F6B0F3A37}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Orleans.Core.Abstractions", "src\Orleans.Core.Abstractions\Orleans.Core.Abstractions.csproj", "{73514686-D25D-478B-9943-A86F6B0F3A37}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "LegacyOrleansLoggerProvider", "src\OrleansLoggerProvider\LegacyOrleansLoggerProvider.csproj", "{0A5009BF-C7D2-463B-BF39-34E1F5EF2F23}"
```
I agree with `Microsoft.Olreans.Logging.Legacy` for root namespace, assembly name and package id, but for directory and project name I would call it `Logging.Legacy` only... The other is too verbose, as every single project in the solution will have the Microsoft.Olreans prefix.

xiazen replied:
@jdom do you agree on the new namespace name?

jdom replied:
yes

xiazen commented on src/OrleansLoggerProvider/EventBulkingDecoratorLogger.cs:
```diff
@@ -0,0 +1,134 @@
+    /// <summary>
+    /// Config for event bulking feature
+    /// </summary>
+    public class EventBulkingConfig
```
sure

xiazen commented on src/OrleansLoggerProvider/EventBulkingDecoratorLogger.cs:
```diff
@@ -0,0 +1,134 @@
+            else
+            {
+                recentLogMessageCounts.AddOrUpdate(logCode, 1, (key, value) => ++value);
+                recentLogMessageCounts.TryGetValue(logCode, out count);
```
will do

xiazen commented on src/OrleansLoggerProvider/EventBulkingDecoratorLogger.cs:
```diff
@@ -0,0 +1,134 @@
+            {
+                // Take local copy of pending bulk message counts, now that this bulk message compaction period has finished
+                var bulkMessageCounts = recentLogMessageCounts.Where(keyPair => keyPair.Value >= this.eventBulkingConfig.BulkEventLimit).ToList();
+                recentLogMessageCounts.Clear();
```
we are supposed to loosing items event when their value is below BulkEventLImit

xiazen commented on src/OrleansLoggerProvider/EventBulkingDecoratorLogger.cs:
```diff
@@ -0,0 +1,134 @@
+                Interlocked.Exchange(ref this.lastBulkLogMessageFlushTicks, now.Ticks);
+          
+                // Output any pending bulk compaction messages
+                if (bulkMessageCounts != null)
```
will refactor

xiazen commented on src/OrleansLoggerProvider/LegacyFileLogConsumer.cs:
```diff
@@ -0,0 +1,82 @@
+
+        public LegacyFileLogConsumer(string fileName)
+        {
+            logOutput = new StreamWriter(File.Open(fileName, FileMode.Append, FileAccess.Write, FileShare.Write));
```
why reading? who is gonna read it? what should we do if multiple thread is trying to log ?

xiazen commented on src/OrleansLoggerProvider/LegacyFileLogConsumer.cs:
```diff
@@ -0,0 +1,82 @@
+            {
+                var msg = string.Format("Ignoring error closing log file {0} - {1}", this.logFileName,
+                    LogFormatter.PrintException(exc));
+                Console.WriteLine(msg);
```
this is trying to preserve old behavior. But sure , I think we can remove this. not enssential

xiazen replied:
hmmm, ConsoleText is internal class, which I think is only designed to use internally? but this csproj are not supposed to have dependency on Orleans , including its internal class eventually. So I think we can address this in another PR to solve the unavailability in AzureIIRC issue

xiazen commented on src/OrleansLoggerProvider/EventBulkingDecoratorLogger.cs:
```diff
@@ -0,0 +1,134 @@
+        /// <summary>
+        /// Default bulk event interval
+        /// </summary>
+        public static readonly TimeSpan DefaultBulkEventInterval = TimeSpan.FromMinutes(1);
```
I think the user will want to know what's the default value , which is why they are public. And we are doing this with our other config classes

xiazen commented on src/OrleansLoggerProvider/EventBulkingDecoratorLogger.cs:
```diff
@@ -0,0 +1,134 @@
+        {
+            var now = DateTime.UtcNow;
+            int count;
+            bool isExcluded = excludedBulkLogCodes.Contains(logCode)
```
sure

xiazen commented on src/OrleansLoggerProvider/LegacyFileLogConsumer.cs:
```diff
@@ -0,0 +1,82 @@
+            int eventCode = 0
+        )
+        {
+            var logMessage = OrleansLoggingUtils.FormatLogMessageToLegacyStyle(DateTime.UtcNow, severity, loggerType, caller, message, ipEndPoint, exception, eventCode, true);
```
it is a util class user can also use to achieve legacy Orleans logging style, which is why it is public

xiazen commented on src/OrleansLoggerProvider/LegacyFileLogConsumer.cs:
```diff
@@ -0,0 +1,82 @@
+        {
+            lock (this.lockObj)
+            {
+                if (this.lockObj == null) return;
```
when one thread closed this consumer , while another thread want to flush it.

xiazen commented:
I address feedbacks

xiazen commented on Directory.Build.props:
```diff
@@ -59,12 +59,13 @@
+    <MicrosoftExtensionsConfigurationJsonVersion>2.0.0</MicrosoftExtensionsConfigurationJsonVersion>
+    <MicrosoftExtensionsConfigurationVersion>2.0.0</MicrosoftExtensionsConfigurationVersion>
+    <MicrosoftExtensionsOptionsVersion>2.0.0</MicrosoftExtensionsOptionsVersion>
+    <MicrosoftExtensionsOptionsConfigurationExtensionsVersion>2.0.0</MicrosoftExtensionsOptionsConfigurationExtensionsVersion>
```
why is that? can you explain more

attilah replied:
Based on our discussions, leave these in the shared section, but make sure to remove the other 2.0s from the current, since they become shared versions and not overrides.

jdom commented:
@dotnet-bot test this please

xiazen commented:
are we changing namespace  to Orleans.Logging.Legacy, based on our 2.0 discussion? 
I think we can change it to that to be consistent with the direction we are currently going. If we are to change the namespace naming principles in the future, we can change this one as well.

attilah commented:
I talked with @jdom about this, he suggested this, I'm not a fan of removing parts so save "screen estate", but we can do it if we decide to do so...

Folder: Logging.Legacy
Project filename: Logging.Legacy.csproj
Namespace: Microsoft.Orleans.Logging.Legacy
Assembly name: Microsoft.Orleans.Logging.Legacy
Package name: Microsoft.Orleans.Logging.Legacy

xiazen commented:
the current PR stick with your suggestion. so we are in an agreement

jdom commented on src/Logging.Legacy/OrleansLoggingUtils.cs:
```diff
@@ -0,0 +1,55 @@
+    public class OrleansLoggingUtils
+    {
+        internal const string ObsoleteMessageString =
+                "The Microsoft.Orleans.Logging.Legacy namespace was kept to facilitate migration from Orleans 1.x but will be removed in the near future. It is recommended that you use the Microsoft.Extensions.Logging infrastructure and providers directly instead of Microsoft.Orleans.Logging.Legacy.Logger and Microsoft.Orleans.Logging.Legacy.ILogConsumer";
```
nit: it's not `Microsoft.Orleans.Logging.Legacy.Logger` any more, it's `Microsoft.Orleans.Logging.Legacy.OrleansLegacyLogger`.

xiazen replied:
? Microsoft.Orleans.Logging.Legacy.Logger is the namespace, which is what we want to refer here right?

xiazen replied:
no I think we are referring to our old Orleans.Runtime.Logger abstraction here. and in the future, Orleans.Runtime.Logger will be changing to Microsoft.Orleans.Logging.Legacy.Logger

jdom replied:
ahh, ok, I remember now :)

jdom commented on src/Logging.Legacy/Logging.Legacy.csproj:
```diff
@@ -0,0 +1,22 @@
+    <PackageId>Microsoft.Orleans.Logging.Legacy</PackageId>
+    <Title>Microsoft Orleans Legacy Logging</Title>
+    <Description>Library of Orleans Legacy Logging.</Description>
+    <PackageTags>$(PackageTags) Orleans</PackageTags>
```
replace with `<PackageTags>$(PackageTags) Logging Legacy</PackageTags>`

jdom commented on src/Logging.Legacy/Logging.Legacy.csproj:
```diff
@@ -0,0 +1,22 @@
+  <PropertyGroup>
+    <PackageId>Microsoft.Orleans.Logging.Legacy</PackageId>
+    <Title>Microsoft Orleans Legacy Logging</Title>
+    <Description>Library of Orleans Legacy Logging.</Description>
```
Please have something more descriptive, as this will go into the nuget package metadata. For example:
This package contains some legacy types used for logging to facilitate migration from Orleans 1.x but will be removed in the near future. It is recommended that you use the Microsoft.Extensions.Logging package and its infrastructure and associated providers directly if not upgrading any legacy log consumer.

jdom commented:
LGTM other than those minor comments
